mixer/adapter/stackdriver/trace/trace.go

```diff
@@ -0,0 +1,197 @@
+
+// Package trace contains a tracespan adapter for Stackdriver trace.
+package trace
+
+import (
+	"context"
+	"fmt"
+
+	"go.opencensus.io/plugin/ochttp"
+	"go.opencensus.io/plugin/ochttp/propagation/b3"
+	"go.opencensus.io/trace"
+
+	"istio.io/istio/mixer/adapter/stackdriver/config"
+	"istio.io/istio/mixer/adapter/stackdriver/helper"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/tracespan"
+)
+
+type (
+	builder struct {
+		types map[string]*tracespan.Type
+		mg    *helper.MetadataGenerator
+		cfg   *config.Params
+	}
+
+	handler struct {
+		te      trace.Exporter
+		sampler trace.Sampler
+	}
+)
+
+var (
+	// compile-time assertion that we implement the interfaces we promise
+	_ tracespan.HandlerBuilder = &builder{}
+	_ tracespan.Handler        = &handler{}
+)
+
+// NewBuilder returns a builder implementing the tracespan.HandlerBuilder interface.
+func NewBuilder(mg *helper.MetadataGenerator) tracespan.HandlerBuilder {
+	return &builder{mg: mg}
+}
+
+func (b *builder) SetTraceSpanTypes(types map[string]*tracespan.Type) {
+	b.types = types
+}
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.cfg = cfg.(*config.Params)
+}
+func (b *builder) Validate() *adapter.ConfigErrors {
+	return nil
+}
+
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+	exporter, err := getExporterFunc(ctx, env, b.cfg)
+	if err != nil {
+		_ = env.Logger().Errorf("Error constructing OpenCensus exporter: %v", err)
+		return nil, err
+	}
+
+	h := &handler{
+		te: exporter,
+	}
+	traceCfg := b.cfg.Trace
+	if sampleProbability := traceCfg.SampleProbability; sampleProbability > 0 {
+		h.sampler = trace.ProbabilitySampler(traceCfg.SampleProbability)
+	}
+	return h, nil
+}
+
+func (h *handler) HandleTraceSpan(_ context.Context, values []*tracespan.Instance) (retErr error) {
+	if h.sampler == nil {
+		// Tracing is not configured.
+		return nil
+	}
+
+	numExported := 0
+	for _, val := range values {
+		parentContext, ok := extractParentContext(val)
+		if !ok {
+			continue
+		}
+		spanContext, ok := extractSpanContext(val, parentContext)
+		if !ok {
+			continue
+		}
+
+		decision := h.sampler(trace.SamplingParameters{
+			ParentContext:   parentContext,
+			TraceID:         spanContext.TraceID,
+			SpanID:          spanContext.SpanID,
+			Name:            val.SpanName,
+			HasRemoteParent: true,
+		})
+
+		if !decision.Sample {
+			continue
+		}
+		spanContext.TraceOptions = trace.TraceOptions(1 /*sampled*/)
+
+		span := buildSpanData(val, parentContext, spanContext)
+		h.te.ExportSpan(span)
+		numExported++
+	}
+
+	if numExported > 0 {
+		h.tryFlush()
+	}
+
+	return
+}
+
+func extractParentContext(val *tracespan.Instance) (trace.SpanContext, bool) {
+	var (
+		parentContext trace.SpanContext
+		ok            bool
+	)
+	if parentContext.TraceID, ok = b3.ParseTraceID(val.TraceId); !ok {
+		return trace.SpanContext{}, false
+	}
+	parentContext.SpanID, _ = b3.ParseSpanID(val.ParentSpanId)
```

Comment:
curious: what are the implications of ignoring the parent span id error ?

should we be generating traceIDs in the cases when they are not present?

Reply:
It doesn't return an error, it returns a `bool` to indicate whether a value was parsed. In this case, we don't care if a value was parsed or not (since parent span ID is optional).